include/circt/Dialect/Emit/EmitOps.td

```diff
@@ -0,0 +1,113 @@
+//
+// This describes the MLIR ops for `emit`.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_EMIT_EMITOPS_TD
+#define CIRCT_DIALECT_EMIT_EMITOPS_TD
+
+include "circt/Dialect/HW/HWTypes.td"
+include "circt/Dialect/Emit/EmitDialect.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+class EmitOp<string mnemonic, list<Trait> traits = []> :
+    Op<EmitDialect, mnemonic, traits>;
+
+def FileOp : EmitOp<"file", [
+    Symbol,
+    SingleBlock,
+    NoTerminator,
+    NoRegionArguments,
+    IsolatedFromAbove
+]> {
+  let summary = "Represents the contents of an emitted file";
+
+  let description = [{
+    This operation groups a set of nested operations to be emitted to a file.
+
+    Other operations (such as file lists)can reference a file to access its
+    filename through an optional symbol.
+  }];
+
+  let regions = (region SizedRegion<1>:$body);
+  let arguments = (ins
+    StrAttr:$file_name,
+    OptionalAttr<SymbolNameAttr>:$sym_name
+  );
+  let results = (outs);
+
+  let assemblyFormat = "$file_name (`sym` $sym_name^)? $body attr-dict";
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    // Creates a file with a callback. While the callback is executed, the
+    // insertion point of the builder is moved inside the body of the file op.
+    OpBuilder<(ins "StringRef":$symName, "StringRef":$fileName,
+                    CArg<"llvm::function_ref<void()>">:$bodyCtor)>,
+    OpBuilder<(ins "StringRef":$fileName,
+                    CArg<"llvm::function_ref<void()>">:$bodyCtor)>,
+  ];
+
+  let extraClassDeclaration = [{
+    // SymbolOpInterface
+    static bool isOptionalSymbol() { return true; }
+
+    // Utilities
+    Block *getBodyBlock() { return &getBodyRegion().front(); }
+  }];
+}
+
+def VerbatimOp : EmitOp<"verbatim"> {
+  let summary = "Verbatim opaque text emitted inline.";
+  let description = [{
+    This operation produces opaque text inline in the file.
+
+    `emit.verbatim` allows symbol reference substitutions with {{0}} syntax.
+  }];
+
+  let arguments = (ins
+    StrAttr:$text
+  );
+
+  let assemblyFormat = [{
+    $text attr-dict
+  }];
+}
+
+
+def FileListOp : EmitOp<"file_list", [
+    Symbol,
+    DeclareOpInterfaceMethods<SymbolUserOpInterface>
+]> {
+  let summary = "Represents a file list";
+
+  let description = [{
+    This operation emits a file list referencing a set of files.
+
+    File lists can be references from other ops (including other file lists)
+    through an optional symbol.
+  }];
+
+  let arguments = (ins
+    StrAttr:$file_name,
+    FlatSymbolRefArrayAttr:$files,
+    OptionalAttr<SymbolNameAttr>:$sym_name
+  );
+  let results = (outs);
+
+  let assemblyFormat = "$file_name `,` $files (`sym` $sym_name^)? attr-dict";
+
+  let extraClassDeclaration = [{
+    // SymbolOpInterface
+    static bool isOptionalSymbol() { return true; }
+  }];
+}
```

Comment:
I think it's fine to proceed with this filelist op. There is an alternative design point which deserves some thought/commentary, though:

The filelist op is, itself a file op. It just has some special handling of how it will emit its `FlatSymbolRefArrayAttr:$file` argument. Put differently, there is an alternative formulation of this that, instead of using a filelist op, puts _filename ops_ inside a file op and treats that as the filelist. This conserves existing concepts more (there is only one op which becomes a file), however, it does not decrease the total number of ops.

This is hinting at the different design points around who knows how to serialize a symbol? Does the op with the symbol know? Does the op that references the symbol know?

WDYT?

Reply:
Once we build out emission a bit it's going to be easier to build up the file lists. Subsequently, we could keep a builder and build a file op out of it.

However, I think that in its current form the op simplifies implementation and even testing, so it might be worth keeping.